lib/spack/spack/test/cmd/env.py

```diff
@@ -149,6 +150,88 @@ def test_env_install_all(install_mockery, mock_fetch):
     assert spec.package.installed
 
 
+def test_env_install_all_fails(install_mockery, mock_fetch, monkeypatch,
+                               capfd):
+    """Test install_all when one of the packages repeatedly fails.
+
+    This test confirms the environment's install_all will proceed after a
+    previous spec fails.  In this case the dependency and dependent are
+    provided, in order, so the dependency will fail once when built explicitly
+    and, again, when it is a dependency.
+    """
+    err_msg = 'Mock _install_task error'
+
+    # Save original _install_task function for conditional use in monkeypatch
+    orig_fn = spack.installer.PackageInstaller._install_task
+
+    def _inst_task(inst, task, **kwargs):
+        spec = task.pkg.spec
+        if spec.name == 'b':
+            raise spack.installer.InstallError(err_msg)
+
+        orig_fn(inst, task, **kwargs)
+
+    monkeypatch.setattr(spack.installer.PackageInstaller, '_install_task',
+                        _inst_task)
+
+    e = ev.create('test-fails')
+    e.add('b')   # attempts to install dependency explicitly
+    e.add('a')   # attempts to install dependent (a) after dependency (b)
+    e.concretize()
+    e.install_all()
+
+    out = capfd.readouterr()[1]
+    env_specs = e._get_environment_specs()
+    for spec in env_specs:
+        assert not spec.package.installed
+    assert out.count(err_msg) == 1  # b_explicit, not as b_dependency
+    assert 'Skipping build of a' in out
+
+
+def test_env_install_all_seq(install_mockery, mock_fetch, monkeypatch):
+    """Test install_all when a successfully installed package is repeatedly
+    installed.
+
+    This test uses the environment installation process to exercise the
+    distributed build multiple times with overlapping dependencies in the
+    same process to ensure proper management of package installation statuses.
+
+    A key part of this test addresses a problem that occurred when attempting
+    to install a package dependent on another successfully installed, multiply
+    referenced package.  An infinite loop would arise after an earlier package
```

Comment:
Based on our discussion I think the issue looks like installing DAGs that overlap, for example

```
X
|
Z
```

followed by 

```
Y
|
Z
```

where both of these DAGs are listed as separate entries in an environment spack.yaml file, and you install all the individual specs in the environment.

Reply:
> Based on our discussion I think the issue looks like installing DAGs that overlap, for example
> 
> ```
> X
> |
> Z
> ```
> 
> followed by
> 
> ```
> Y
> |
> Z
> ```
> 
> where both of these DAGs are listed as separate entries in an environment spack.yaml file, and you install all the individual specs in the environment.

You are correct that the diamond pattern is not the simplest case where this can occur.  I was trying to re-use existing packages.  I added the example below to the test code.  Given packages `A`, `B`, and `Depb`, where `A` and `Depb` depend on `B`:

```
    A      Depb
    |       |
    B       B
```

If the installation of `B` succeeds while `A` fails, then the installation of `Depb` should succeed.